### PR TITLE
Create SearchFilter on work packages 

### PR DESCRIPTION
**Pull request by wielinde**
Searches on
- subject
- description
- attachment file name
- attachment file content
- comments

https://community.openproject.com/wp/29116

Out of scope for now:
- [ ] Search on custom fields type text

### CODE REVIEW

**Comment by lindenthal** [comment]
hi @wielinde,

why do you restrict the search on global custom fields?

Best
Niels

**Comment by wielinde** [comment]
@lindenthal I didn't know it better back than. I removed the restriction. However, I deprioritized it to "Nice to have" as I don't have the impression that the feature was strongly desired. Please feel free to give it a priority boost.

File: app/models/queries/work_packages/filter/comment_filter.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class Queries::WorkPackages::Filter::CommentFilter < Queries::WorkPackages::Filter::WorkPackageFilter
+  def type
+    :text
+  end
+
+  def includes
+    %i{project journals}
```
Comment (ulferts) [inline app/models/queries/work_packages/filter/comment_filter.rb]: Why does `project` need to be included?

File: app/models/queries/work_packages/filter/search_filter.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  CONTAINS_OPERATOR = '~'.freeze
+
+  def filters
+    if @filters
+      @filters.each do |filter|
+        filter.operator = CONTAINS_OPERATOR
+        filter.values = values
+      end
+    else
+      filter_class_list = [
```
Comment (ulferts) [inline app/models/queries/work_packages/filter/search_filter.rb]: The handling of what filters are eligible should be extracted into a separate method to shorten the `filters` method.

File: app/models/queries/work_packages/filter/search_filter.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        filter_class_list += [
+          [Queries::WorkPackages::Filter::AttachmentContentFilter, :attachment_content],
+          [Queries::WorkPackages::Filter::AttachmentFileNameFilter, :attachment_file_name]
+        ]
+      end
+      @filters = filter_class_list.map do |filter_class|
+        filter_class.first.create!(name: filter_class.second,
+                                   context: context,
+                                   operator: CONTAINS_OPERATOR,
+                                   values: values)
```
Comment (ulferts) [inline app/models/queries/work_packages/filter/search_filter.rb]:
```
@filters = filter_class_list.map do |filter_class, filter_name|
  filter_class.create!(name: filter_name,
                       context: context,
                       operator: CONTAINS_OPERATOR,
                       values: values)
```
is equivalent and easier to read

File: spec/factories/attachment_factory.rb [hunk 4]
```diff
@@ -52,5 +52,7 @@
     factory :attached_picture do
       content_type { 'image/jpeg' }
     end
+
+
```
Comment (ulferts) [inline spec/factories/attachment_factory.rb]: remove, please

**Review by ulferts — CHANGES_REQUESTED** [review]
Functionality looks good, please see my few remarks within the code.

**Comment by wielinde** [comment]
@ulferts Thank you for your review. I took it as an opportunity to extract an OrBaseFilter that can serve us as a base for an IdAndStatusFilter for the Autocompleter, too.

File: app/models/queries/work_packages/filter/search_filter.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    I18n.t('label_search')
+  end
+
+  def includes
+    filters.map(&:includes).flatten.uniq.reject(&:blank?)
+  end
+
+  def filter_configurations
+    list = CE_FILTERS
+    list + EE_TSV_FILTERS if EnterpriseToken.allows_to?(:attachment_filters) && OpenProject::Database.allows_tsv?
```
Comment (ulferts) [inline app/models/queries/work_packages/filter/search_filter.rb]: `nil` will be returned from the method if the condition is not met which is not expected by the code

File: app/models/queries/work_packages/filter/or_base_filter.rb [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        filter.values = values
+      end
+    else
+      @filter = create_instances
+    end
+  end
+
+  def self.key
+    raise NotImplementedError
+  end
```
Comment (ulferts) [inline app/models/queries/work_packages/filter/or_base_filter.rb]: I think this should be removed. Doing that will mean that the functionality will fall back to a sensible default defined in   `Queries::Filters::Base`. That way, you wouldn't have to define the method again in `Queries::WorkPackages::Filter::SearchFilter`

File: app/models/queries/work_packages/filter/or_base_filter.rb [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+  def name
+    raise NotImplementedError
+  end
+
+  def type
+    raise NotImplementedError
+  end
```
Comment (ulferts) [inline app/models/queries/work_packages/filter/or_base_filter.rb]:
```
def self.type
```
is already defined in the same way in Queries::Filters::Base

File: app/models/queries/work_packages/filter/or_base_filter.rb [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    end
+  end
+
+  def self.key
+    raise NotImplementedError
+  end
+
+  def name
+    raise NotImplementedError
+  end
```
Comment (ulferts) [inline app/models/queries/work_packages/filter/or_base_filter.rb]: Is this actually required? The base filter already defines the behaviour, so if a subclass wants to rely on that it should be able to.

File: app/models/queries/work_packages/filter/or_base_filter.rb [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+  def type
+    raise NotImplementedError
+  end
+
+  def human_name
+    raise NotImplementedError
+  end
```
Comment (ulferts) [inline app/models/queries/work_packages/filter/or_base_filter.rb]: Again, defined in base class.

File: app/models/queries/work_packages/filter/or_base_filter.rb [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+  def update_instances
+    configurations = filter_configurations
+
+    @filters.each_with_index do |filter, index|
+      operator_for_instance = configurations[index].third
+
+      filter.operator = operator_for_instance
+      filter.values = values
+    end
+  end
```
Comment (ulferts) [inline app/models/queries/work_packages/filter/or_base_filter.rb]: This method is defined, but not used in e.g. https://github.com/opf/openproject/pull/6879/files#diff-34c9d42ca6cd72148d28262066292d1dR34

File: app/models/queries/work_packages/filter/or_base_filter.rb [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+                           operator: operator,
+                           values: values)
+    end
+  end
+
+  def update_instances
+    configurations = filter_configurations
+
+    @filters.each_with_index do |filter, index|
+      operator_for_instance = configurations[index].third
```
Comment (ulferts) [inline app/models/queries/work_packages/filter/or_base_filter.rb]: You rely on the subclass knowing that it has to provide an array of arrays (with at least three elements each) in `filter_configurations` and to know the semantics of each index in the inner arrays. To increase stability, please introduce a class/struct or at least turn it into an array of hashes.

File: app/models/queries/work_packages/filter/or_base_filter.rb [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class Queries::WorkPackages::Filter::OrBaseFilter < Queries::WorkPackages::Filter::WorkPackageFilter
```
Comment (ulferts) [inline app/models/queries/work_packages/filter/or_base_filter.rb]: From the looks of it, using a mixin instead of a class would fit the intention better.

**Review by ulferts — CHANGES_REQUESTED** [review]
I like the idea of extracting the or functionality.

**Pull request merged**